README.md
**alpaka** - Abstraction Library for Parallel Kernel Acceleration
=================================================================

This is a CI test only.

[![Continuous Integration](https://github.com/alpaka-group/alpaka/workflows/Continuous%20Integration/badge.svg)](https://github.com/alpaka-group/alpaka/actions?query=workflow%3A%22Continuous+Integration%22)
[![Documentation Status](https://readthedocs.org/projects/alpaka/badge/?version=latest)](https://alpaka.readthedocs.io)
[![Doxygen](https://img.shields.io/badge/API-Doxygen-blue.svg)](https://alpaka-group.github.io/alpaka)
[![Language](https://img.shields.io/badge/language-C%2B%2B17-orange.svg)](https://isocpp.org/)
[![Platforms](https://img.shields.io/badge/platform-linux%20%7C%20windows%20%7C%20mac-lightgrey.svg)](https://github.com/alpaka-group/alpaka)
[![License](https://img.shields.io/badge/license-MPL--2.0-blue.svg)](https://www.mozilla.org/en-US/MPL/2.0/)

![alpaka](docs/logo/alpaka_401x135.png)

The **alpaka** library is a header-only C++17 abstraction library for accelerator development.

Its aim is to provide performance portability across accelerators through the abstraction (not hiding!) of the underlying levels of parallelism.

It is platform independent and supports the concurrent and cooperative use of multiple devices such as the hosts CPU as well as attached accelerators as for instance CUDA GPUs and Xeon Phis (currently native execution only).
A multitude of accelerator back-end variants using CUDA, OpenMP (2.0/5.0), std::thread and also serial execution is provided and can be selected depending on the device.
Only one implementation of the user kernel is required by representing them as function objects with a special interface.
There is no need to write special CUDA, OpenMP or custom threading code.
Accelerator back-ends can be mixed within a device queue.
The decision which accelerator back-end executes which kernel can be made at runtime.

The abstraction used is very similar to the CUDA grid-blocks-threads division strategy.
Algorithms that should be parallelized have to be divided into a multi-dimensional grid consisting of small uniform work items.
These functions are called kernels and are executed in parallel threads.
The threads in the grid are organized in blocks.
All threads in a block are executed in parallel and can interact via fast shared memory.
Blocks are executed independently and can not interact in any way.
The block execution order is unspecified and depends on the accelerator in use.
By using this abstraction the execution can be optimally adapted to the available hardware.


Software License
----------------

**alpaka** is licensed under **MPL-2.0**.


Documentation
-------------

The alpaka documentation can be found in the [online manual](https://alpaka.readthedocs.io).
The documentation files in [`.rst` (reStructuredText)](https://www.sphinx-doc.org/en/stable/rest.html) format are located in the `docs` subfolder of this repository.
The [source code documentation](https://alpaka-group.github.io/alpaka/) is generated with [doxygen](http://www.doxygen.org).


Accelerator Back-ends
---------------------

| Accelerator Back-end   | Lib/API                                                 | Devices                | Execution strategy grid-blocks     | Execution strategy block-threads     |
|------------------------|---------------------------------------------------------|------------------------|------------------------------------|--------------------------------------|
| Serial                 | n/a                                                     | Host CPU (single core) | sequential                         | sequential (only 1 thread per block) |
| OpenMP 2.0+ blocks     | OpenMP 2.0+                                             | Host CPU (multi core)  | parallel (preemptive multitasking) | sequential (only 1 thread per block) |
| OpenMP 2.0+ threads    | OpenMP 2.0+                                             | Host CPU (multi core)  | sequential                         | parallel (preemptive multitasking)   |
| std::thread            | std::thread                                             | Host CPU (multi core)  | sequential                         | parallel (preemptive multitasking)   |
| TBB                    | TBB 2.2+                                                | Host CPU (multi core)  | parallel (preemptive multitasking) | sequential (only 1 thread per block) |
| CUDA                   | CUDA 9.0+                                               | NVIDIA GPUs            | parallel (undefined)               | parallel (lock-step within warps)    |
| HIP(clang)             | [HIP 5.0+](https://github.com/ROCm-Developer-Tools/HIP) | AMD GPUs               | parallel (undefined)               | parallel (lock-step within warps)    |


Supported Compilers
-------------------

This library uses C++17 (or newer when available).

| Accelerator Back-end                                                           | gcc 9.5 <br/> (Linux)                       | gcc 10.4 / 11.2 <br/> (Linux)               | gcc 12.2 <br/> (Linux)                      | clang 9 <br/> (Linux)                                  | clang 10 / 11<br/> (Linux)                              | clang 12 <br/> (Linux)                           | clang 13 <br/> (Linux)                      | clang 14 <br/> (Linux)                            | clang 15 <br/> (Linux)                | clang 16 <br/> (Linux)       | icpx (latest)      | Xcode 13.2.1 / 14.2 <br /> (macOS)                    | Visual Studio 2022 <br/> (Windows)   |
|--------------------------------------------------------------------------------|---------------------------------------------|---------------------------------------------|---------------------------------------------|--------------------------------------------------------|---------------------------------------------------------|--------------------------------------------------|---------------------------------------------|---------------------------------------------------|---------------------------------------|------------------------------|---------------------|-------------------------------------------------------|--------------------------------------|
| Serial                                                                         | :white_check_mark:                          | :white_check_mark:                          | :white_check_mark:                          | :white_check_mark:                                     | :white_check_mark:                                      | :white_check_mark:                               | :white_check_mark:                          | :white_check_mark:                                | :white_check_mark:                    | -                            | :white_check_mark: | :white_check_mark:                                    | :white_check_mark:                   |
| OpenMP 2.0+ blocks                                                             | :white_check_mark:                          | :white_check_mark:                          | :white_check_mark:                          | :white_check_mark:                                     | :white_check_mark:                                      | :white_check_mark:                               | :white_check_mark:                          | :white_check_mark:                                | :white_check_mark:                    | -                            | :white_check_mark: | :white_check_mark:                                    | :white_check_mark:                   |
| OpenMP 2.0+ threads                                                            | :white_check_mark:                          | :white_check_mark:                          | :white_check_mark:                          | :white_check_mark:                                     | :white_check_mark:                                      | :white_check_mark:                               | :white_check_mark:                          | :white_check_mark:                                | :white_check_mark:                    | -                            | :white_check_mark: | :white_check_mark:                                    | :white_check_mark:                   |
| std::thread                                                                    | :white_check_mark:                          | :white_check_mark:                          | :white_check_mark:                          | :white_check_mark:                                     | :white_check_mark:                                      | :white_check_mark:                               | :white_check_mark:                          | :white_check_mark:                                | :white_check_mark:                    | -                            | :white_check_mark: | :white_check_mark:                                    | :white_check_mark:                   |
| TBB                                                                            | :white_check_mark:                          | :white_check_mark:                          | :white_check_mark:                          | :white_check_mark:                                     | :white_check_mark:                                      | :white_check_mark:                               | :white_check_mark:                          | :white_check_mark:                                | :white_check_mark:                    | -                            | :white_check_mark: | :white_check_mark:                                    | :white_check_mark:                   |
| CUDA (nvcc)                                                                    | :white_check_mark: <br/> (CUDA 11.0 - 12.0) | :white_check_mark: <br/> (CUDA 11.4 - 12.0) | :white_check_mark: <br/> (CUDA 12.0 - 12.1) | :white_check_mark: <br/> (CUDA 11.0-11.2; 11.6 - 12.0) | :white_check_mark: <br/> (CUDA 11.2, 11.6 - 12.0)       | :white_check_mark: <br/> (CUDA 11.6 - 12.0)      | :white_check_mark: <br/> (CUDA 11.7 - 12.0) | :white_check_mark: <br/> (CUDA 11.8 - 12.0)       | :white_check_mark: <br/> (CUDA 12.1)  | -                            | :x:                | :x:                                                   | :x:                                  |
| CUDA (clang)                                                                   | -                                           | -                                           | -                                           | :x:                                                    | :x:                                                     | :x:                                              | :x:                                         | :white_check_mark: (CUDA 11.0 - 11.5)             | :white_check_mark: (CUDA 11.0 - 11.5) | -                            | :x:                | -                                                     | -                                    |
| [HIP](https://alpaka.readthedocs.io/en/latest/install/HIP.html) (clang)        | -                                           | -                                           | -                                           | :x:                                                    | :x:                                                     | :x:                                              | :x:                                         | :white_check_mark: (HIP 5.0 - 5.2)                | :white_check_mark: (HIP 5.3 - 5.4)    | :white_check_mark: (HIP 5.5) | :x:                | -                                                     | -                                    |

Other compilers or combinations marked with :x: in the table above may work but are not tested in CI and are therefore not explicitly supported.

Dependencies
------------

[Boost](https://boost.org/) 1.74.0+ is the only mandatory external dependency.
The **alpaka** library itself just requires header-only libraries.
However some of the accelerator back-end implementations require different boost libraries to be built.

When an accelerator back-end using *CUDA* is enabled, version *11.0* (with nvcc as CUDA compiler) or version *9.2* (with clang as CUDA compiler) of the *CUDA SDK* is the minimum requirement.
*NOTE*: When using clang as a native *CUDA* compiler, the *CUDA accelerator back-end* can not be enabled together with any *OpenMP accelerator back-end* because this combination is currently unsupported.
*NOTE*: Separable compilation is disabled by default and can be enabled via the CMake flag `CMAKE_CUDA_SEPARABLE_COMPILATION`.

When an accelerator back-end using *OpenMP* is enabled, the compiler and the platform have to support the corresponding minimum *OpenMP* version.

When an accelerator back-end using *TBB* is enabled, the compiler and the platform have to support the corresponding minimum *TBB* version.


Usage
-----

The library is header only so nothing has to be built.
CMake 3.22+ is required to provide the correct defines and include paths.
Just call `alpaka_add_executable` instead of `add_executable` and the difficulties of the CUDA nvcc compiler in handling `.cu` and `.cpp` files are automatically taken care of.
Source files do not need any special file ending.
Examples of how to utilize alpaka within CMake can be found in the `example` folder.

The whole alpaka library can be included with: `#include <alpaka/alpaka.hpp>`
Code that is not intended to be utilized by the user is hidden in the `detail` namespace.

Furthermore, for a CUDA-like experience when adopting alpaka we provide the library [*cupla*](https://github.com/alpaka-group/cupla).
It enables a simple and straightforward way of porting existing CUDA applications to alpaka and thus to a variety of accelerators.

Introduction
------------

For a quick introduction, feel free to playback the recording of our presentation at
[GTC 2016](https://www.nvidia.com/gtc/):

 - E. Zenker, R. Widera, G. Juckeland et al.,
   *Porting the Plasma Simulation PIConGPU to Heterogeneous Architectures with Alpaka*,
   [video link (39 min)](http://on-demand.gputechconf.com/gtc/2016/video/S6298.html),
   [slides (PDF)](https://on-demand.gputechconf.com/gtc/2016/presentation/s6298-erik-zenker-porting-the-plasma.pdf),
   [DOI:10.5281/zenodo.6336086](https://doi.org/10.5281/zenodo.6336086)


Citing alpaka
-------------

Currently all authors of **alpaka** are scientists or connected with
research. For us to justify the importance and impact of our work, please
consider citing us accordingly in your derived work and publications:

```latex
% Peer-Reviewed Publication %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% Peer reviewed and accepted publication in
%   "2nd International Workshop on Performance Portable
%    Programming Models for Accelerators (P^3MA)"
% colocated with the
%   "2017 ISC High Performance Conference"
%   in Frankfurt, Germany
@inproceedings{MathesP3MA2017,
  author    = {{Matthes}, A. and {Widera}, R. and {Zenker}, E. and {Worpitz}, B. and
               {Huebl}, A. and {Bussmann}, M.},
  title     = {Tuning and optimization for a variety of many-core architectures without changing a single line of implementation code
               using the Alpaka library},
  archivePrefix = "arXiv",
  eprint    = {1706.10086},
  keywords  = {Computer Science - Distributed, Parallel, and Cluster Computing},
  day       = {30},
  month     = {Jun},
  year      = {2017},
  url       = {https://arxiv.org/abs/1706.10086},
}

% Peer-Reviewed Publication %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% Peer reviewed and accepted publication in
%   "The Sixth International Workshop on
%    Accelerators and Hybrid Exascale Systems (AsHES)"
% at the
%   "30th IEEE International Parallel and Distributed
%    Processing Symposium" in Chicago, IL, USA
@inproceedings{ZenkerAsHES2016,
  author    = {Erik Zenker and Benjamin Worpitz and Ren{\'{e}} Widera
               and Axel Huebl and Guido Juckeland and
               Andreas Kn{\"{u}}pfer and Wolfgang E. Nagel and Michael Bussmann},
  title     = {Alpaka - An Abstraction Library for Parallel Kernel Acceleration},
  archivePrefix = "arXiv",
  eprint    = {1602.08477},
  keywords  = {Computer science;CUDA;Mathematical Software;nVidia;OpenMP;Package;
               performance portability;Portability;Tesla K20;Tesla K80},
  day       = {23},
  month     = {May},
  year      = {2016},
  publisher = {IEEE Computer Society},
  url       = {http://arxiv.org/abs/1602.08477},
}


% Original Work: Benjamin Worpitz' Master Thesis %%%%%%%%%%
%
@MasterThesis{Worpitz2015,
  author = {Benjamin Worpitz},
  title  = {Investigating performance portability of a highly scalable
            particle-in-cell simulation code on various multi-core
            architectures},
  school = {{Technische Universit{\"{a}}t Dresden}},
  month  = {Sep},
  year   = {2015},
  type   = {Master Thesis},
  doi    = {10.5281/zenodo.49768},
  url    = {http://dx.doi.org/10.5281/zenodo.49768}
}
```

Contributing
------------

Rules for contributions can be found in [CONTRIBUTING.md](CONTRIBUTING.md)

Authors
-------

### Maintainers* and Core Developers

- Benjamin Worpitz* (original author)
- Dr. Sergei Bastrakov*
- Dr. Andrea Bocci
- Dr. Antonio Di Pilato
- Simeon Ehrig
- Bernhard Manfred Gruber*
- Dr. Axel Huebl
- Dr. Jeffrey Kelling
- Dr. Felice Pantaleo
- Jan Stephan*
- Dr. Jiří Vyskočil
- René Widera*

### Former Members, Contributions and Thanks

- Dr. Michael Bussmann
- Mat Colgrove
- Valentin Gehrke
- Maximilian Knespel
- Jakob Krude
- Alexander Matthes
- Hauke Mewes
- Phil Nash
- Dr. David M. Rogers
- Mutsuo Saito
- Jonas Schenke
- Daniel Vollmer
- Matthias Werner
- Bert Wesarg
- Malte Zacharias
- Erik Zenker


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [0.6.1] - 2021-XX-XX
### Compatibility Changes:
- rework implementation of OpenMP schedule support #1279 #1309
  - `alpaka::omp::Schedule` is replaced by `ompScheduleKind` and `ompScheduleChunkSize`
### Bug Fixes:
- CI: use ubuntu-18.04 for gcc-5 and gcc-6 builds #1252
- fix OpenMP 5 shared memory allocation #1254 
- fix static shared memory alignment #1282
- fix BlockSharedMemStMemberImpl::getVarPtr for last var #1280
- fix CPU static shared memory implementation #1258
- unit tests: fix queue test #1266
- fix CtxBlockOacc: SyncBlockThreads #1291
- port macOSX CI fix from #1283
- fix assert in DeclareSharedVar (OpenAcc) #1303
- CI: disable GCC 10.3 + NVCC tests #1302
- example: fix warning (NVCC+OpenMP) #1307
- CMake CUDA: dev compile options not propagated #1294

### Misc
- add ALPAKA_ASSERT_OFFLOAD Macro #1260
- document return value of `empty()` and `isComplete()` #1265


## [0.6.0] - 2021-01-20
### Compatibility Changes:
- support for CUDA 11, 11.1, and 11.2 #1076 #1086 #1147 #1231
- remove support for CUDA 11.0 with MSVC 2019 #1227
- support for CMake 3.18.0 and 3.19.0 #1087 #1217
- set minimal HIP version to 3.5 #1110
- remove CMake HIP module shipped with alpaka #1189
- set HIP-clang as default compiler for HIP #1113
- support for NVCC + VS 2019 #1121
- support for boost-1.74.0 #1142
- explicitly require backends and do not enable them by default #1111
- remove support for Xcode 11.1 #1206 
- support Xcode 11.21 - 12.2.0 #1206
- update to Catch 2.13.3 #1215

### Bug Fixes:
- apply some clang-tidy fixes #1044
- fix CUDA/HIP accelerator concept usage #1064
- fix Intel compiler detection #1070
- CMake: build type CXX flag not passed to nvcc #1073
- work around Intel ICE (Internal Compiler Error) when using std::decay on empty template parameter packs #1074
- BoostPredef.hpp: Add redefinition of BOOST_COMP_PGI #1082
- fix min/max return type deduction #1085
- CMake: fix boost fiber linking #1088
- fix HIP-clang compile #1107
- fix CUDA/HIP cmake flags #1152
- fix error handling CUDA/HIP #1108
- ALPAKA_DECAY_T: Fix Intel detection, Add PGI #1116
- fix how to set HIP target architecture #1112 
- fix and improve block shared mem st member sanity checks #1128
- HIP: remove copy device2device workaround #1188
- pass native pointers to kernel instead of buffer objects #1193
- fix bug in `isPinned()` and `pin()` #1196
- fix marking of unit tests for concepts #1226

### New Features:
- add functions `alpaka::atomicAnd` et. al. as shortcuts to `alpaka::atomicOp<alpaka::AtomicAnd>` et. al. #1005
- warp voting functions #1003 #1049 #1090 #1092
- Sphinx Doc: Fix Doxygen integration on readthedocs #1042 #1093 #1151
- add cheat sheet to the docs #1057 #1177
- extend AccDevProps with shared memory size per block #1084
- OpenMP 5 target offload backend #1126
- OpenACC backend #1127
- option to set OpenMP schedule for the Omp2Blocks backend #1223

### Misc
- tests for BufferSlicing #1024
- use std::invoke_result_t instead of std::result_of_t when available #1047
- simplify shared memory usage in tests #1075 
- remove boost::aligned_alloc #1094
- add unit tests for work div #1095
- change examples (except reduce) to use getValidWorkDiv #1104
- example monte-carlo-integration #1106 
- invoke docker run only once instead of twice #1109
- cpu/SysInfo.hpp: Add #else for cpuid; Add PGI #1119
- Pgi std atomic workaround #1120
- make BlockSharedMemDynMember::staticAllocBytes a function #1118
- add IntrinsicFallback: basic fallback implementations #1122
- allow ALPAKA_CXX_STANDARD to propagate to nvcc with MSVC 1920 and above #1130
- add set kernel #1132
- make Queue test generic to handle QueueGenericThreads* with different devices #1133
- IdxBtOmp: Add GetIdx specialization for 1d #1140
- test CMAKE_CXX_EXTENSIONS=OFF #1153
- change block memory size back to be stored as 32 bit #1187
- add comments to math function traits that explain valid argument range #1190
- provide docker_retry #1191
- add .clang-format file #1204
- add CI check whether code is correctly formatted #1213
- make test/common a CMake INTERFACE library #1228

### Breaking changes:

The namespace structure of *alpaka* is now flattened. 
The [script](https://gist.github.com/sliwowitz/0a55e1bed6350f7fcae17ef0d430040d) can help you to apply the changes to your code.
The script only works if you used the full namespace `alpaka::*` for alpaka functions.

- removed namespace `alpaka::dev`
- removed namespace `alpaka::pltf`
- renamed function `alpaka::vec::cast` to `alpaka::castVec`
- renamed function `alpaka::vec::reverse` to `alpaka::reverseVec`
- renamed function `alpaka::vec::concat` to `alpaka::concatVec`
- removed namespace `alpaka::vec`
- removed namespace `alpaka::workdiv`
- removed namespace `alpaka::acc`
- renamed functors `alpaka::atomic::op::And` et. al. to `alpaka::AtomicAnd` et. al. #1185
- removed namespace `alpaka::atomic::op`
- removed namespace `alpaka::atomic`
- removed namespace `alpaka::queue`
- removed namespace `alpaka::idx`
- removed namespace `alpaka::dim`
- removed namespace `alpaka::kernel`
- removed namespace `alpaka::wait`
- removed namespace `alpaka::mem`
- removed namespace `alpaka::offset`
- removed namespace `alpaka::elem`
- removed namespace `alpaka::intrinsic`
- renamed function `alpaka::event::test` to `alpaka::isComplete`
- removed namespace `alpaka::event`
- removed namespace `alpaka::time`
- removed namespace `alpaka::example`
- renamed function `alpaka::alloc::alloc` to `alpaka::malloc`
- renamed function `alpaka::buf::alloc` to `alpaka::allocBuf`
- removed namespace `alpaka::alloc`
- removed namespace `alpaka::buf`
- renamed function `alpaka::view::set` to `alpaka::memset`
- renamed function `alpaka::view::copy` to `alpaka::memcpy`
- removed namespace `alpaka::view`
- removed namespace `alpaka::block::shared::st`

script/install_cuda.sh
#!/bin/bash

#
# Copyright 2017-2019 Benjamin Worpitz
#
# This file is part of alpaka.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#

source ./script/travis_retry.sh

source ./script/set.sh

: "${ALPAKA_CUDA_VERSION?'ALPAKA_CUDA_VERSION must be specified'}"

ALPAKA_CUDA_VER_SEMANTIC=( ${ALPAKA_CUDA_VERSION//./ } )
ALPAKA_CUDA_VER_MAJOR="${ALPAKA_CUDA_VER_SEMANTIC[0]}"
echo ALPAKA_CUDA_VER_MAJOR: "${ALPAKA_CUDA_VER_MAJOR}"

if [ "$ALPAKA_CI_OS_NAME" = "Linux" ]
then
    : "${ALPAKA_CI_CUDA_DIR?'ALPAKA_CI_CUDA_DIR must be specified'}"
    : "${ALPAKA_CUDA_COMPILER?'ALPAKA_CUDA_COMPILER must be specified'}"

    # Ubuntu 18.04 requires some extra keys for verification
    if [[ "$(cat /etc/os-release)" == *"18.04"* ]]
    then
        travis_retry sudo apt-get -y --quiet --allow-unauthenticated --no-install-recommends install dirmngr gpg-agent
        travis_retry sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys F60F4B3D7FA2AF80
    elif [[ "$(cat /etc/os-release)" == *"20.04"* ]]
    then
        travis_retry sudo apt-get -y --quiet --allow-unauthenticated --no-install-recommends install dirmngr gpg-agent
        travis_retry sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys F60F4B3D7FA2AF80
    fi

    # Set the correct CUDA downloads
    if [ "${ALPAKA_CUDA_VERSION}" == "9.0" ]
    then
        ALPAKA_CUDA_PKG_DEB_NAME=cuda-repo-ubuntu1604-9-0-local
        ALPAKA_CUDA_PKG_FILE_NAME="${ALPAKA_CUDA_PKG_DEB_NAME}"_9.0.176-1_amd64-deb
        ALPAKA_CUDA_PKG_FILE_PATH=https://developer.nvidia.com/compute/cuda/9.0/Prod/local_installers/${ALPAKA_CUDA_PKG_FILE_NAME}
    elif [ "${ALPAKA_CUDA_VERSION}" == "9.1" ]
    then
        ALPAKA_CUDA_PKG_DEB_NAME=cuda-repo-ubuntu1604-9-1-local
        ALPAKA_CUDA_PKG_FILE_NAME="${ALPAKA_CUDA_PKG_DEB_NAME}"_9.1.85-1_amd64
        ALPAKA_CUDA_PKG_FILE_PATH=https://developer.nvidia.com/compute/cuda/9.1/Prod/local_installers/${ALPAKA_CUDA_PKG_FILE_NAME}
    elif [ "${ALPAKA_CUDA_VERSION}" == "9.2" ]
    then
        ALPAKA_CUDA_PKG_DEB_NAME=cuda-repo-ubuntu1604-9-2-local
        ALPAKA_CUDA_PKG_FILE_NAME="${ALPAKA_CUDA_PKG_DEB_NAME}"_9.2.88-1_amd64
        ALPAKA_CUDA_PKG_FILE_PATH=https://developer.nvidia.com/compute/cuda/9.2/Prod/local_installers/${ALPAKA_CUDA_PKG_FILE_NAME}
    elif [ "${ALPAKA_CUDA_VERSION}" == "10.0" ]
    then
        ALPAKA_CUDA_PKG_DEB_NAME=cuda-repo-ubuntu1804-10-0-local
        ALPAKA_CUDA_PKG_FILE_NAME="${ALPAKA_CUDA_PKG_DEB_NAME}"-10.0.130-410.48_1.0-1_amd64
        ALPAKA_CUDA_PKG_FILE_PATH=https://developer.nvidia.com/compute/cuda/10.0/Prod/local_installers/${ALPAKA_CUDA_PKG_FILE_NAME}
    elif [ "${ALPAKA_CUDA_VERSION}" == "10.1" ]
    then
        ALPAKA_CUDA_PKG_DEB_NAME=cuda-repo-ubuntu1804-10-1-local
        ALPAKA_CUDA_PKG_FILE_NAME="${ALPAKA_CUDA_PKG_DEB_NAME}"-10.1.168-418.67_1.0-1_amd64.deb
        ALPAKA_CUDA_PKG_FILE_PATH=https://developer.nvidia.com/compute/cuda/10.1/Prod/local_installers/${ALPAKA_CUDA_PKG_FILE_NAME}
    elif [ "${ALPAKA_CUDA_VERSION}" == "10.2" ]
    then
        ALPAKA_CUDA_PKG_DEB_NAME=cuda-repo-ubuntu1804-10-2-local
        ALPAKA_CUDA_PKG_FILE_NAME="${ALPAKA_CUDA_PKG_DEB_NAME}"-10.2.89-440.33.01_1.0-1_amd64.deb
        ALPAKA_CUDA_PKG_FILE_PATH=http://developer.download.nvidia.com/compute/cuda/10.2/Prod/local_installers/${ALPAKA_CUDA_PKG_FILE_NAME}
    elif [ "${ALPAKA_CUDA_VERSION}" == "11.0" ]
    then
        ALPAKA_CUDA_PKG_DEB_NAME=cuda-repo-ubuntu1804-11-0-local
        ALPAKA_CUDA_PKG_FILE_NAME="${ALPAKA_CUDA_PKG_DEB_NAME}"_11.0.3-450.51.06-1_amd64.deb
        ALPAKA_CUDA_PKG_FILE_PATH=http://developer.download.nvidia.com/compute/cuda/11.0.3/local_installers/${ALPAKA_CUDA_PKG_FILE_NAME}
    elif [ "${ALPAKA_CUDA_VERSION}" == "11.1" ]
    then
        ALPAKA_CUDA_PKG_DEB_NAME=cuda-repo-ubuntu1804-11-1-local
        ALPAKA_CUDA_PKG_FILE_NAME="${ALPAKA_CUDA_PKG_DEB_NAME}"_11.1.0-455.23.05-1_amd64.deb
        ALPAKA_CUDA_PKG_FILE_PATH=http://developer.download.nvidia.com/compute/cuda/11.1.0/local_installers/${ALPAKA_CUDA_PKG_FILE_NAME}
    else
        echo CUDA versions other than 9.0, 9.1, 9.2, 10.0, 10.1, 10.2, 11.0 and 11.1 are not currently supported on linux!
    fi
    if [ -z "$(ls -A ${ALPAKA_CI_CUDA_DIR})" ]
    then
        mkdir -p "${ALPAKA_CI_CUDA_DIR}"
        travis_retry wget --no-verbose -O "${ALPAKA_CI_CUDA_DIR}"/"${ALPAKA_CUDA_PKG_FILE_NAME}" "${ALPAKA_CUDA_PKG_FILE_PATH}"
    fi
    sudo dpkg --install "${ALPAKA_CI_CUDA_DIR}"/"${ALPAKA_CUDA_PKG_FILE_NAME}"

    travis_retry sudo apt-get -y --quiet update

    # Install CUDA
    # Currently we do not install CUDA fully: sudo apt-get --quiet -y install cuda
    # We only install the minimal packages. Because of our manual partial installation we have to create a symlink at /usr/local/cuda
    if (( "${ALPAKA_CUDA_VER_MAJOR}" >= 11 ))
    then
      sudo apt-get -y --quiet --allow-unauthenticated --no-install-recommends install cuda-compiler-"${ALPAKA_CUDA_VERSION}" cuda-cudart-"${ALPAKA_CUDA_VERSION}" cuda-cudart-dev-"${ALPAKA_CUDA_VERSION}" libcurand-"${ALPAKA_CUDA_VERSION}" libcurand-dev-"${ALPAKA_CUDA_VERSION}"
    else
      sudo apt-get -y --quiet --allow-unauthenticated --no-install-recommends install cuda-core-"${ALPAKA_CUDA_VERSION}" cuda-cudart-"${ALPAKA_CUDA_VERSION}" cuda-cudart-dev-"${ALPAKA_CUDA_VERSION}" cuda-curand-"${ALPAKA_CUDA_VERSION}" cuda-curand-dev-"${ALPAKA_CUDA_VERSION}"
    fi
    sudo ln -s /usr/local/cuda-"${ALPAKA_CUDA_VERSION}" /usr/local/cuda

    if [ "${ALPAKA_CUDA_COMPILER}" == "clang" ]
    then
        travis_retry sudo apt-get -y --quiet --allow-unauthenticated --no-install-recommends install g++-multilib
    fi

    # clean up
    sudo rm -rf "${ALPAKA_CI_CUDA_DIR}"/"${ALPAKA_CUDA_PKG_FILE_NAME}"
    sudo dpkg --purge "${ALPAKA_CUDA_PKG_DEB_NAME}"
elif [ "$ALPAKA_CI_OS_NAME" = "Windows" ]
then
    if [ "${ALPAKA_CUDA_VERSION}" == "10.0" ]
    then
        ALPAKA_CUDA_PKG_FILE_NAME=cuda_10.0.130_411.31_win10
        ALPAKA_CUDA_PKG_FILE_PATH=https://developer.nvidia.com/compute/cuda/10.0/Prod/local_installers/${ALPAKA_CUDA_PKG_FILE_NAME}
    elif [ "${ALPAKA_CUDA_VERSION}" == "10.1" ]
    then
        ALPAKA_CUDA_PKG_FILE_NAME=cuda_10.1.168_425.25_win10.exe
        ALPAKA_CUDA_PKG_FILE_PATH=https://developer.nvidia.com/compute/cuda/10.1/Prod/local_installers/${ALPAKA_CUDA_PKG_FILE_NAME}
    elif [ "${ALPAKA_CUDA_VERSION}" == "10.2" ]
    then
        ALPAKA_CUDA_PKG_FILE_NAME=cuda_10.2.89_441.22_win10.exe
        ALPAKA_CUDA_PKG_FILE_PATH=http://developer.download.nvidia.com/compute/cuda/10.2/Prod/local_installers/${ALPAKA_CUDA_PKG_FILE_NAME}
    elif [ "${ALPAKA_CUDA_VERSION}" == "11.0" ]
    then
        ALPAKA_CUDA_PKG_FILE_NAME=cuda_11.0.2_451.48_win10.exe
        ALPAKA_CUDA_PKG_FILE_PATH=http://developer.download.nvidia.com/compute/cuda/11.0.2/local_installers/${ALPAKA_CUDA_PKG_FILE_NAME}
    elif [ "${ALPAKA_CUDA_VERSION}" == "11.1" ]
    then
        ALPAKA_CUDA_PKG_FILE_NAME=cuda_11.1.0_456.43_win10.exe
        ALPAKA_CUDA_PKG_FILE_PATH=http://developer.download.nvidia.com/compute/cuda/11.1.0/local_installers/${ALPAKA_CUDA_PKG_FILE_NAME}
    else
        echo CUDA versions other than 10.0, 10.1, 10.2, 11.0 and 11.1 are not currently supported on Windows!
    fi

    curl -L -o cuda_installer.exe ${ALPAKA_CUDA_PKG_FILE_PATH}
    ./cuda_installer.exe -s "nvcc_${ALPAKA_CUDA_VERSION}" "curand_dev_${ALPAKA_CUDA_VERSION}" "cudart_${ALPAKA_CUDA_VERSION}"
fi
